File: 0094. Binary Tree Inorder Traversal/Solution.cpp
// Author        : tarlansoltanov
// Title         : Binary Tree Inorder Traversal
// ID            : 0094
// URL           : https://leetcode.com/problems/binary-tree-inorder-traversal/
// Difficulty    : Easy
// Language      : C++


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        if(root == nullptr) return {};
        
        vector<int> res = {}, l, r;
        
        if(root->left != nullptr){
            l = inorderTraversal(root->left);
            res.insert(res.end(), l.begin(), l.end());
        }

        res.push_back(root->val);

        if(root->right != nullptr){
            r = inorderTraversal(root->right);
            res.insert(res.end(), r.begin(), r.end());
        }

        return res;
    }
};

File: 0094. Binary Tree Inorder Traversal/Solution.py
# Author        : tarlansoltanov
# Title         : Binary Tree Inorder Traversal
# ID            : 0094
# URL           : https://leetcode.com/problems/binary-tree-inorder-traversal/
# Difficulty    : Easy
# Language      : Python


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        res = []

        def explore(node: Optional[TreeNode]):
            if node is None:
                return
            explore(node.left)
            res.append(node.val)
            explore(node.right)

        explore(root)
        return res

